[Uploading models guide] Update docs (#1266)
huggingface/hub-docs: docs/hub/models-uploading.md

You can link repositories with an individual user, such as [osanseviero/fashion_brands_patterns](https://huggingface.co/osanseviero/fashion_brands_patterns), or with an organization, such as [facebook/bart-large-xsum](https://huggingface.co/facebook/bart-large-xsum). Organizations can collect models related to a company, community, or library! If you choose an organization, the model will be featured on the organization’s page, and every member of the organization will have the ability to contribute to the repository. You can create a new organization [here](https://huggingface.co/organizations/new).

> > **_NOTE:_** Models do NOT need to be compatible with the Transformers/Diffusers libraries to get download metrics. Any custom PyTorch model is supported. Read more below!

Reviewer: This read too pytorch-centric. People can use sklearn/TF/etc

```suggestion
> **_NOTE:_** Models do NOT need to be compatible with the Transformers/Diffusers libraries to get download metrics. Any custom model is supported. Read more below!
```

> There are several ways to upload models to the Hub and get download metrics, described below.

Reviewer 2: ```suggestion
There are several ways to upload models for them to be nicely integrated into the Hub and get download metrics, described below.
```

- In case your model is designed for a library that has [built-in support](#upload-from-a-library-with-built-in-support), you can use the methods provided by the library. Custom models that use `trust_remote_code=True` can also leverage these methods.
- In case your model is a custom PyTorch model, one can leverage the [`PyTorchModelHubMixin` class](#upload-a-pytorch-model-using-huggingface_hub) as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module` class, just like models in the Transformers, Diffusers and Timm libraries.
- In addition to programmatic uploads, you can always use the [web interface](#using-the-web-interface).

> Do note that only the first two ways will ensure [download metrics](models-download-stats) work for your model.

Reviewer: Although people can also do their own files and update model-libraries.ts, as https://github.com/huggingface/huggingface.js/pull/561/files

Reviewer 2: i would just remove the focus on download numbers from this doc page, and just link to the [relevant doc page](https://huggingface.co/docs/hub/models-download-stats)

We can still mention the PytorchMixin

> Do note that only the first two ways will ensure [download metrics](models-download-stats) work for your model.

Reviewer 2: ```suggestion
```

There are several ways to upload models to the Hub, described below.

- In case your model comes from a library that has [built-in support](#upload-from-a-library-with-built-in-support), one can use the existing methods.

> - In case your model is a custom PyTorch model, the two recommended ways to integrate with the ecosystem is to either:

Reviewer: WDYT about adding a bolded message in the first paragraph clarifying that repos don't need to be `transformers`-based

Author: Great tip, added it!

>     - use the `transformers`'s [custom code](https://huggingface.co/docs/transformers/custom_models) feature, which makes your model directly usable in the Transformers library

Reviewer 2: this is a bit too transformers specific at this stage no?

Reviewer 2: the rest LGTM from a quick read

Reviewer: Agree, and I see this more of a `model comes from a library that has ...` class. E.g. we also have remote code with `diffusers`.

> - In case your model is a custom PyTorch model, there are 2 ways:
>     - using the [custom code](https://huggingface.co/docs/transformers/custom_models) feature on the hub, which makes your model usable in the Transformers library
>     - leveraging the [huggingface_hub](#upload-a-pytorch-model-using-huggingface_hub) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module`, just like models in the Transformers, Diffusers and Timm libraries.

Reviewer 3: ```suggestion
- In case your model is a custom PyTorch model, the two recommended ways to integrate with the ecosystem is to either:
    - use the `transformers`'s [custom code](https://huggingface.co/docs/transformers/custom_models) feature, which makes your model directly usable in the Transformers library
    - or leverage the [`huggingface_hub`'s `PyTorchModelHubMixin` class](#upload-a-pytorch-model-using-huggingface_hub) as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module` class, just like models in the Transformers, Diffusers and Timm libraries.
```

- In addition to programmatic uploads, you can always use the [web interface](#using-the-web-interface).

> Do note that only the first 2 ways will make sure that download metrics work for your model.

Reviewer: ```suggestion
Do note that only the first two ways will ensure download metrics work for your model.
```

Reviewer: Maybe add a link to https://huggingface.co/docs/hub/models-download-stats

> Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model and make it easier discoverable.

Reviewer 3: ```suggestion
Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model and make it more discoverable.
```

class MyModel(nn.Module, PyTorchModelHubMixin):
    def __init__(self, num_channels: int, hidden_size: int, num_classes: int):
        super().__init__()
        self.param = nn.Parameter(torch.rand(num_channels, hidden_size))
        self.linear = nn.Linear(hidden_size, num_classes)

    def forward(self, x):
        return self.linear(x + self.param)

# create model
config = {"num_channels": 3, "hidden_size": 32, "num_classes": 10}
model = MyModel(**config)

# save locally
model.save_pretrained("my-awesome-model")

# push to the hub
model.push_to_hub("my-awesome-model")

# reload
model = MyModel.from_pretrained("username/my-awesome-model")
```

> As can be seen, the only thing required is to inherit from `PyTorchModelHubMixin`. All attributes will be automatically serialized as a `config.json`.

Reviewer 3: ```suggestion
As can be seen, the only thing required is to inherit from `PyTorchModelHubMixin`. All attributes will be automatically serialized in a `config.json` file.
```
